Documentation/wip/testing-proposal.md

```diff
@@ -0,0 +1,66 @@
+
+```
+./fission
+./fission_old
+```
+
+1. Fission_old is built and deployed in the cluster and then tests are run. This is exact same as the current build and test workflow but for a previous version of Fission.
+
+2. Once (1) passes, script switches to fission directory and builds and pushes the images. The part where the workflow diverts is install_and_test, and here are the differences:
+
+- Instead of ```helm install```, we will run ```helm upgrade``` - and upgrade the existing cluster with newer images
+- Script will use the new fission CLI
+- Script will use tests in fission_old to run the tests
+
+3. The script for (1) can be used from existing codebase while the part (2) will need developing a part of script specifically for upgrade.
+
+4. After the run - if both set of tests succeed, the upgrade can be considered fairly successful (Although not entirely - as there can be potential gap)
+
+### Alternative approach considered
+
+It is possible to get the binary from previous release instead of source code - but it has two drawbacks:
+
+1. The effort for (2) in previous section still remains the same
+2. With binary install, we loose the tests from older version of Fission.
+
+## A small Diversion
+
+While researching for the best way to run upgrade tests, I was looking at testing frameworks used by other go projects. I found Ginkgo (A BDD framework) and Gomega  (A matcher library). The experessiveness of the tests and constructs for common issues such as retry-backoff etc. are well built and I found some aspects which are relevant to our tests. Details below:
+
+### Issues & possible solution
+
+- Most of our tests are sequential - but in real life there are multiple things going on a server and we should mimic that - i.e. run tests in parallel. This is also benefetial to reduce overall test time. (Of course there are exceptions)
+
+Ginkgo allows tests to run in parallel and fires them in separate threads. It also allows randomizing the tests for every execution
+
+
+- Run selective test suites with certain purpose, for ex. test all performance tests is very rudimentry  with shell scripts. Also it is difficult to specify some some actions which should be run "before executing test suite 1" and after finishing execution and so on.
+
+Ginkgo provides constructs such as ```BeforeSuite``` and ```AfterSuite``` and the same for each test with ```BeforeEach`` and ```AfterEach```. This allows clear separation of fixtures from actual tests and also common setup and teardown
+
+- Some of our flaky tests succeed sometime and fail sometime. A lot of time is spent in debugging and fixinng in PRs on those tests. While the test itself should be fixed in longer term, it should be possible to retry flaky test as part of execution. Ginkgo provides a mechanism to retry certain tests.
```

Comment:
> Some of our flaky tests succeed sometime and fail sometime. A lot of time is spent in debugging and fixinng in PRs on those tests. While the test itself should be fixed in longer term, it should be possible to retry flaky test as part of execution.

A few small grammatical errors. Maybe instead:

*Sometimes, some of our tests fail unexpectedly and pass on the next attempt. A lot of time is spent in debugging and fixing in PRs on those tests. While each test should be fixed eventually, it should be possible to retry a flaky test as part of execution.*